### ISSUE TITLE
Normalize organisations names

### ISSUE THREAD

**Issue by vitorbaptista** [opentrials/opentrials#604]
Organisation names are tricky: you can go as far as looking for the ownership trees to find companies with different names that have the same owner. This issue isn't about that, but just dealing with the simplest case: organisations with the same name but different wordings (e.g. `ACME` and `ACME Ltda`). They should all have the same name in our database. In general, prefer the longer wording (e.g. `Johnson and Johnson` instead of `J&J`).

This will probably involve changes on https://github.com/opentrials/processors/blob/master/processors/base/writers/organisation.py, so all processors using this writer will automatically have its name cleaned.

**Comment by vitorbaptista** [comment]
https://github.com/datamade/dedupe and https://discuss.okfn.org/t/entity-reconciliation-services/1785/5 might be helpful here

**Comment by felipevieira** [comment]
+1 for Dedupe lib. Seems the way to go

**Comment by felipevieira** [comment]
I'm currently thinking about moving forward with a single solution that initially computes some organization clusters from scratch using all the organization data present on the database. Afterwards, we can use these clusters to search for one that contains the org entry we are trying to normalize: if we find it, the longest member of the cluster is returned, otherwise we just add this new entry to the cluster data, re-cluster (so we can assign - or not - this new entry to a specific cluster) and go back to the searching phase. 

Any thoughts or suggestions about this approach?

**Comment by vitorbaptista** [comment]
@felipevieira These clusters would be processed offline and then used by the processors? In other words, it would be similar to @arthurSena's solution for trial phase normalisation, but instead of creating the clusters manually, they would be created automatically?

Sounds sensible. Just remember that, if we're using a file to save the clusters, they need to get into the git repository. It might be better to create a table in the `warehouse` database with the clusters, so we can update it as we go.

**Comment by felipevieira** [comment]
@vitorbaptista Currently I'm processing these clusters when a processor starts (before the registers processing starts, to avoid useless reprocessing at each register), this way there's no need of any auxiliar file containing the organisations mapping, as I can keep everything I need stored in memory. However, since this pre-processing is taking a considerable amount of time (~7s) because we have more than 11k organisations stored on the database, I'm really planning to perform this pre-processing offline, as you've mentioned. 

Do you also think that this cluster processing shouldn't be attached to the processor? Besides, what would be a good approach to allow these clusters to be updated periodically offline? A scheduled task?

**Comment by felipevieira** [comment]
Hi everyone. What goes next is the current approach I'm using to normalize the organisations (the content in **bold** reffers to new features). I have some questions about the process and any thoughts on that will be very helpful:

1 - **If user has explicitly requested to update the organisatons clusters (through a `make start --recompute-clusters` flag)**
    &nbsp;&nbsp;&nbsp;&nbsp;1.1 - **Fetch all organisations stored on database**
    &nbsp;&nbsp;&nbsp;&nbsp;1.2 - **Using dedupe, build organisations clusters according to the orgs fetched on 1.1 and store these clusters on the warehouse database for further offline access**
2 - For each register on the source database of the warehouse
   &nbsp;&nbsp;&nbsp;&nbsp; 2.1 - **If organisation is already stored on database**
        &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.1.1 - **If organisation is allocated in any stored cluster, return the cluster canonical form (the longest one) and proceed to the usual organisation storage and trial linking processes**
       &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2.1.2 - **Otherwise return the organisation itself (if organisation isn't clustered, that's because it's already its unique and canonical form) and proceed to the usual organisation storage and linking processes**
   &nbsp;&nbsp;&nbsp;&nbsp; 2.2 - **Otherwise, ?**

The main question here is what have to be done in 2.2. It's intuitive to see that, if  a organisation is new (not stored on database) we have to check if it can be clustered with any of the existent forms and then return it's canonical form or itself if it continues unclustered. However, since we have a large amount of organisations on database, the clustering process is currently taking ~7s (at least on my local machine), so in the worst case (where most of the organisations of the registers are new) we will be adding an overhead of 7s for _each processed register_, so it doesn't seem a very good approach. (let's call it approach 1). The approach 2 would be a manual clustering (without using dedup to recluster) where we would check each cluster stored and check if I can add my new org to any cluster according to some threshold. That would avoid the reclustering overhead, but we haven't found so far an effective way to compare strings in the same way dedupe does (especially avoiding to add new depedencies), so the current comparison approaches would lead to incorrect clusterings (for example, using fuzzywuzzy to check string similarity with a high threshold is clustering "" and ""). The last (and imho the simplest) approach to normalize new entries would be also returning the organisation itself (as informed) without any kind of normalization, for one simple reason: that's true that we might have unnormalized organisations for a short interval, but in the next time when user start a processor with the reclustering flag, the clusters would be regenerated and all unnormalized forms previously stored would be properly clustered all at once with the single overhead of one clustering process (and not multiple clustering overheads). With the clusters properly built, these old and incorrect normalized forms would be then fixed (since all registers are always reprocessed).

I apologize for the size of my explanation. I just wanted to both inform the solution I've implemented so far and to describe exactly my questions about how to finish this algorithm. What I'm expecting next is any thoughts about which approach (1, 2 or 3) to follow or even a new alternative to adjust this algorithm to better fit our requirements. Once my questions are answered I can proceed to finish my implementation and send my PR.

**Comment by vitorbaptista** [comment]
@felipevieira Sorry for the delayed response. The approach `3` seems sensible to me. I'd like to understand better the step `2` of your process, though ("For each register on the source database..."). Are you saying that there'll be a processor that'll go over all organisations and fix them, or you're talking about whenever a new processor (say `nct` processor) runs it'll clean the organisation names?

Also, @NickCEBM has been working on a manual two-way table to normalize the organisation names. Nick, could you share the table here? It would be very useful as a way to validate the automated process @felipevieira is working on.

**Comment by felipevieira** [comment]
@vitorbaptista thanks for your thoughts here. Would you please reopen this issue (even though GitHub says I've closed it, I don't recall doing that)?

About the step ``2``, I'm justing reviewing what all processors do currently. If I'm correct, all processors go through every register stored on warehouse (for a specific source, such as ``nct``) and update the trials informations (including organisation). Thus, I can ensure that even though some trials might be assigned to unnormalized organisations at first (because these organisations are new and unclustered), this data will be properly normalized in a next processing (when these new organisations will be clustered). Was I clearer now?

Also, A manual way to validate my results would be great. Looking forward to see this two-way table to check the algorithm accuracy

**Comment by vitorbaptista** [comment]
@felipevieira Yes, that's OK. Also, instead of creating a new `make start --recompute-clusters` flag, simply create a new processor that recomputes them.

**Comment by NickCEBM** [comment]
Hey, so here is the table but a few notes:

- I only normalized on names that appeared more than twice (except for NHS Trusts).
- I made some decisions that likely won't be useful to validate against such as trying to track down acquisitions and in many instances grouping university hospital systems under the umbrella of the university.
- For now, frequent co-sponsors who appeared on the same line were normalized as joint sponsors
- This list only picked one sponsor per study (we just used the output of calling the Max sponsor name in SQL). The full data sometimes has different sponsors listed across country studies, the same sponsor inputted in different ways or multiple sponsors in the same study.

All that being said, for the big, very frequent commercial sponsors this should help out to check against.

[normalized_sponsor_names.xlsx](https://github.com/opentrials/opentrials/files/760200/normalized_sponsor_names.xlsx)

**Comment by felipevieira** [comment]
@vitorbaptista I was taking a look at @NickCEBM's normalized table and realized that it's not according to the longer wording policy you've defined on the issue description (e.g. according to it, ``World Health Organisation`` is normalized to ``WHO``), so I'm not sure if we can use this file to validate our normalization algorithm
